enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequestHandler.java

```diff
@@ -0,0 +1,134 @@
+    {
+        log.debug( "Requesting Lucene snapshot %s", snapshotRequest );
+        CloseablesListener closeablesListener = new CloseablesListener();
+        boolean differentStore = !snapshotRequest.expectedStoreId().equalToKernelStoreId( dataSource.get().getStoreId() );
+        boolean transactionTooFarBehind = !isTransactionWithinReach( snapshotRequest.requiredTransactionId() );
+        if ( differentStore )
+        {
+            storeFileStreamingProtocol.end( ctx, StoreCopyFinishedResponse.Status.E_STORE_ID_MISMATCH );
+            protocol.expect( CatchupServerProtocol.State.MESSAGE_TYPE );
+        }
+        else if ( transactionTooFarBehind )
+        {
+            storeFileStreamingProtocol.end( ctx, StoreCopyFinishedResponse.Status.E_TOO_FAR_BEHIND );
+        }
+        else
+        {
+            File storeDir = dataSource.get().getStoreDir();
+            ResourceIterator<StoreFileMetadata> resourceIterator =
+                    dataSource.get().getNeoStoreIndexListing().getSnapshot( snapshotRequest.descriptor() );
+            try
+            {
+                closeablesListener.sample( resourceIterator );
+                while ( resourceIterator.hasNext() )
+                {
+                    StoreFileMetadata storeFileMetadata = resourceIterator.next();
+                    storeFileStreamingProtocol.stream( ctx,
+                            new StoreResource( storeFileMetadata.file(), relativePath( storeDir, storeFileMetadata.file() ), storeFileMetadata.recordSize(),
+                                    pageCache, fs ) );
+                }
+                storeFileStreamingProtocol.end( ctx, StoreCopyFinishedResponse.Status.SUCCESS ).addListener( closeablesListener );
+            }
+            catch ( Throwable e )
+            {
+                storeFileStreamingProtocol.end( ctx, StoreCopyFinishedResponse.Status.E_UNKNOWN ).addListener( closeablesListener );
+            }
+            finally
+            {
+                protocol.expect( CatchupServerProtocol.State.MESSAGE_TYPE );
+            }
+        }
+    }
+
+    private boolean isTransactionWithinReach( long atLeast )
+    {
+        if ( isEqualOrGreater( atLeast ) )
+        {
+            return true;
+        }
+        else
+        {
+            try
+            {
+                checkpointerSupplier.get().tryCheckPoint( new SimpleTriggerInfo( "Store file copy" ) );
+                return isEqualOrGreater( atLeast );
+            }
+            catch ( IOException e )
+            {
+                log.error( "Unable to acquire lock while preparing to serve catchup, consider the the transaction to be `behind`" );
+                return false;
+            }
+        }
+    }
+
+    private boolean isEqualOrGreater( long atLeast )
+    {
+        return checkpointerSupplier.get().lastCheckPointedTransactionId() >= atLeast;
```

Comment:
If this is false the request will fail and the client will have to retry somewhere else. We do not try to achieve that this is true its only a condition for being able to continue.

Currently all requests are to the leader so this should always be true. But we want to be able to send requests to other instances, possible we will have to force checkpointing to happen to avoid this condition constantly failing?